### ISSUE TITLE
Can we notify Polymer about changes once entire patch applied?

### ISSUE THREAD

**Issue by miyconst** [PuppetJs/puppet-polymer-client#29]
It is very bug prone when Polymer observers are hit in between of patch items.

cc @warpech, @tomalec

**Comment by warpech** [comment]
I'd rather avoid it, because then the client side code becomes aware of Puppet.

Can you give any examples when its really unavoidable?

As a workaround, perhaps you can have a property `Updated: 0` that is incremented on the server and set the observer on it?

**Comment by miyconst** [comment]
>I'd rather avoid it, because then the client side code becomes aware of Puppet.

Just opposite. It should be hidden into the `puppet-polymer-client`. Currently we have to be aware that the view-model can get into inconsistent state at any moment and care about it in every web component.

The last case which raised this issue:

```html
<palma-d3-matrix rows="{{model.Rows}}" cols="{{model.Cols}}" cells="{{model.Cells}}"></palma-d3-matrix>
```

The `Rows`, `Cols`, and `Cells` arrays are updated on certain user action. Now the problem is that `palma-d3-matrix` observes changes on those values and expect to have cells for existing rows and cols. What happens in reality is that polymer notifies `palma-d3-matrix` about `Cells` change before the `Rows` and `Cols` have actually been updated.

Even more weird staff happening when `partial.Html` property is involved.

**Comment by warpech** [comment]
What version of Starcounter and puppet-polymer-client are you using? Current version uses `notifyPath` (see https://github.com/PuppetJs/puppet-polymer-client/blob/gh-pages/puppet-client.html#L467) after all of the sequence operations are actually applied.

**Comment by miyconst** [comment]
You are right. After some more digging and investigation I found that my issue is caused by the following:

- Client receives a patch to replace a partial.
- PuppetJs applies view-model changes.
- `puppet-polymer-client` notifies Polymer about the changes.
- Polymer applies the view-model changes to the **old** HTML.
- Exception happens here, as the old HTML is not designed for the new view-model.
- `starcounter-include` removes old HTML.
- `starcounter-include` loads new partial.

**Comment by miyconst** [comment]
I have created a simplest possible example in my [Playground](https://github.com/miyconst/Playground/).

In the example I have two partials with `{ "Date": "" }` property defined, loading one after another will apply the new value to old HTML first and only then replace the entire partial.

Steps to reproduce:

1. Clone the repository.
2. Start Playground app.
3. Open `http://localhost:8080/playground`.
4. Open Chrome Console.
5. Click `[Load date page]` button.
6. Click `[Load sub page]` button.
7. **The `DatePage` partial will throw an exception because the value from `SubPage` was applied to it.**

Replace [`console.error`](https://github.com/miyconst/Playground/blob/master/Playground/wwwroot/Playground/DatePage.html#L16) with `debugger` if you want to stop for debugging.

**Comment by tomalec** [comment]
For the first look, it seems to me like a bug (missing scenario coverage) in `starcounter-include` (`juicy-html`). Could you provide an example of a patch received by client?

**Comment by miyconst** [comment]
@tomalec here is the patch:

```json
[
  {
    "op": "replace",
    "path": "/_ver#s",
    "value": 3
  },
  {
    "op": "test",
    "path": "/_ver#c$",
    "value": 2
  },
  {
    "op": "replace",
    "path": "/SubPage",
    "value": {
      "Playground": {
        "Html": "/Playground/SubPage.html",
        "Date": "SubPage",
        "Items": [ "i7-4790k", "i7-5820k", "i7-6700k" ],
        "SelectedItem$": "",
        "Details": { },
        "Remove$": 0,
        "Add$": 0,
        "Message": ""
      }
    }
  },
  {
    "op": "replace",
    "path": "/LoadSubPage$",
    "value": 0
  }
]
```

Can you look at my example? It is tiny and easy to understand.

**Comment by warpech** [comment]
Let's look at this when Polymer 2.0 integration is tested in Starcounter 2.4.
